File: M/1705-maximum-number-of-eaten-apples/main.go
package main

import (
	"container/heap"
	"fmt"
)

type MinHeap [][]int

func (p MinHeap) Len() int {
	return len(p)
}

func (p MinHeap) Swap(i, j int) {
	p[i], p[j] = p[j], p[i]
}

func (p MinHeap) Less(i, j int) bool {
	return p[i][1] < p[j][1]
}

func (p *MinHeap) Push(c interface{}) {
	*p = append(*p, c.([]int))
}
func (p *MinHeap) Pop() interface{} {
	n := len(*p)
	v := (*p)[n-1]
	*p = (*p)[0 : n-1]
	return v
}

func eatenApples(apples []int, days []int) int {
	h := &MinHeap{}
	heap.Init(h)

	cnt := 0
	for i := 0; i < len(apples); i++ {
		if apples[i] > 0 {
			heap.Push(h, []int{apples[i], i + days[i]})
		}
		for len(*h) > 0 {
			candi := heap.Pop(h).([]int)
			if candi[1] > i {
				candi[0]--
				if candi[0] > 0 {
					heap.Push(h, candi)
				}
				cnt++
				break
			}
		}
	}

	day := len(apples)
	for len(*h) > 0 {
		candi := heap.Pop(h).([]int)
		if candi[1] > day {
			candi[0]--
			if candi[0] > 0 {
				heap.Push(h, candi)
			}
			cnt++
			day++
		}
	}
	return cnt
}

func main() {
	fmt.Println()
}

File: M/1705-maximum-number-of-eaten-apples/main.rs
struct Solution {}

use std::cmp::Reverse;
use std::collections::BinaryHeap;
impl Solution {
  pub fn eaten_apples(apples: Vec<i32>, days: Vec<i32>) -> i32 {
    let mut bh: BinaryHeap<Reverse<(i32, i32)>> = BinaryHeap::new();
    let mut ans: i32 = 0;

    (0..apples.len()).for_each(|day| {
      if apples[day] > 0 {
        bh.push(Reverse((day as i32 + days[day], apples[day])));
      }
      while bh.len() > 0 {
        let candi = bh.pop().unwrap().0;
        if candi.0 > day as i32 {
          let r = candi.1;
          if r > 1 {
            bh.push(Reverse((candi.0, r - 1)));
          }
          ans += 1;
          break;
        }
      }
    });
    let mut day = apples.len() as i32;
    while bh.len() > 0 {
      let candi = bh.pop().unwrap().0;
      if candi.0 > day {
        let r = candi.1;
        if r > 1 {
          bh.push(Reverse((candi.0, r - 1)));
        }
        ans += 1;
        day += 1;
      }
    }

    ans
  }
}

fn main() {
  println!(
    "{}",
    Solution::eaten_apples(vec![1, 2, 3, 5, 2], vec![3, 2, 1, 4, 2])
  );
  println!(
    "{}",
    Solution::eaten_apples(vec![3, 0, 0, 0, 0, 2], vec![3, 0, 0, 0, 0, 2])
  );
}
